### PR TITLE
254 mintburn audit

### PR DESCRIPTION
**Pull request by sirpy**
# Description

#254

### CODE REVIEW

File: contracts/governance/GoodDollarStaking.sol [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+		public
+		nonReentrant
+		returns (uint256 goodRewards, uint256 gdRewards)
+	{
+		uint256 depositComponent;
+
+		//in case amount is 0 this will just withdraw the GOOD rewards, this is required for withdrawRewards
+		if (_amount > 0) {
+			/* G$ rewards update */
+			//we get the relative part user is withdrawing from his original deposit, his principle is composed of deposit+earned interest
```
Comment (Shloyem) [inline contracts/governance/GoodDollarStaking.sol:139]: Here and on the function comment it says principle, when we use withdraw (that behinds the scenes uses principle so maybe its ok)

File: test/utils/GoodDollarMintBurnWrapper.test.ts [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+      )
+    );
+
+    expect(await goodDollar.balanceOf(founder.address)).to.eq(0); //verify burn happened
+    expect(await goodDollar.balanceOf(wrapper.address)).to.eq(0); //verify burn happened
+    const events = await multiChainRouter.queryFilter(
+      multiChainRouter.filters.AnySwap()
+    );
+    expect(events[0].args.recipient).to.equal(minter.address);
+    expect(events[0].args.chainId).to.equal(4220);
```
Comment (Shloyem) [inline test/utils/GoodDollarMintBurnWrapper.test.ts:657]: can const

File: contracts/utils/GoodDollarMintBurnWrapper.sol [hunk 3]
```diff
@@ -148,32 +132,47 @@ contract GoodDollarMintBurnWrapper is
...
-		MintBurnAny, // mint and burn(address from, uint256 amount), don't need approve
-		MintBurnFrom, // mint and burnFrom(address from, uint256 amount), need approve
-		MintBurnSelf, // mint and burn(uint256 amount), call transferFrom first, need approve
-		Transfer, // transfer and transferFrom, need approve
-		TransferDeposit // transfer and transferFrom, deposit and withdraw, need approve
-	}
-
 	address public token; // the target token this contract is wrapping
-	TokenType public tokenType;
+	uint256 private unused_tokenType;
```
Comment (Shloyem) [inline contracts/utils/GoodDollarMintBurnWrapper.sol:144]: Is this needed?

File: contracts/utils/GoodDollarMintBurnWrapper.sol [hunk 4]
```diff
@@ -300,10 +261,14 @@ contract GoodDollarMintBurnWrapper is
 		return true;
 	}
 
-	// impl IRouter `burn`
+	/**
+	 * @notice implement the IRouter burn required for work with multichain router. This method is used by the multichain bridge to burn tokens on sidechain
+	 * on bridge transfer to other chain. Can only be called by the ROUTER role
+	 * @param from sender - requires sender to first tokens to the wrapper or use transferAndCall
```
Comment (Shloyem) [inline contracts/utils/GoodDollarMintBurnWrapper.sol:267]: to first .. something .. tokens

File: contracts/utils/GoodDollarMintBurnWrapper.sol [hunk 5]
```diff
@@ -372,55 +326,215 @@ contract GoodDollarMintBurnWrapper is
...
+			totalMintDebt -= uint128(toBurn);
+			ERC20(token).burn(toBurn); //from DAOUpgradableContract -> Interfaces
+		}
+	}
+
+	/**
+	 * @notice helper for sendOrMint action to update the rewarder daily limit if updateFrequency passed
+	 */
+	function _updateDailyLimitCap(Supply storage minter) internal {
+		uint256 secondsPased = block.timestamp - minter.lastUpdate;
```
Comment (Shloyem) [inline contracts/utils/GoodDollarMintBurnWrapper.sol:525]:
```suggestion
		uint256 secondsPassed = block.timestamp - minter.lastUpdate;
```

File: contracts/utils/GoodDollarMintBurnWrapper.sol [hunk 6]
```diff
@@ -372,55 +326,215 @@ contract GoodDollarMintBurnWrapper is
...
+
+		if (secondsPased >= updateFrequency) {
```
Comment (Shloyem) [inline contracts/utils/GoodDollarMintBurnWrapper.sol:527]:
```suggestion
		if (secondsPassed >= updateFrequency) {
```

File: contracts/utils/GoodDollarMintBurnWrapper.sol [hunk 7]
```diff
@@ -372,55 +326,215 @@ contract GoodDollarMintBurnWrapper is
...
+			minter.dailyCap = uint128(
+				(IERC20Upgradeable(token).totalSupply() * minter.bpsPerDay) / 10000
+			);
+			minter.lastUpdate = uint128(block.timestamp);
+			// console.log(
+			// 	"secondsPassed %s %s %s",
+			// 	secondsPased,
```
Comment (Shloyem) [inline contracts/utils/GoodDollarMintBurnWrapper.sol:534]:
```suggestion
			// 	secondsPassed,
```

**Review by Shloyem — APPROVED** [review]

**Review by MehrdadBadriGol — APPROVED** [review]

**Pull request merged**